ccl/changefeedccl: TestChangefeedContinuousTelemetryOnTermination failed

ccl/changefeedccl.TestChangefeedContinuousTelemetryOnTermination [failed](https://tanzanite.cluster.engflow.com/invocations/default/4b4caeb2-0964-44f6-81ae-b79407262409?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on release-25.1 @ [899f1e1a73780f9d12d1a4a61a5f82f2eb4ead94](https://github.com/cockroachdb/cockroach/commits/899f1e1a73780f9d12d1a4a61a5f82f2eb4ead94):


```
=== RUN   TestChangefeedContinuousTelemetryOnTermination
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestChangefeedContinuousTelemetryOnTermination1644168868
    test_log_scope.go:76: use -show-logs to present logs inline
    changefeed_test.go:6411: -- test log scope end --
test logs left over in: outputs.zip/logTestChangefeedContinuousTelemetryOnTermination1644168868
--- FAIL: TestChangefeedContinuousTelemetryOnTermination (18.17s)
=== RUN   TestChangefeedContinuousTelemetryOnTermination/cloudstorage
    helpers_test.go:883: making server as secondary tenant
    helpers_test.go:962: making cloudstorage feed factory
    testfeed_test.go:301: creating external connection
    testfeed_test.go:304: ran create external connection
    helpers_test.go:1280: read message {{{1743329161533029297 changefeed_emitted_bytes} ‹CREATE CHANGEFEED FOR TABLE foo INTO 'external://testconn2199349612120145408' WITH OPTIONS (format = 'parquet')› external 1 no yes parquet 1059254308694753281} 637 1 86400000000000 false}
    helpers_test.go:1280: read message {{{1743329166801677541 changefeed_emitted_bytes} ‹CREATE CHANGEFEED FOR TABLE foo INTO 'external://testconn2199349612120145408' WITH OPTIONS (format = 'parquet')› external 1 no yes parquet 1059254308694753281} 0 0 86400000000000 true}
    helpers_test.go:1280: read message {{{1743329166777610387 changefeed_emitted_bytes} ‹CREATE CHANGEFEED FOR TABLE foo INTO 'external://testconn2199349612120145408' WITH OPTIONS (format = 'parquet')› external 1 no yes parquet 1059254308694753281} 645 2 86400000000000 false}
    helpers_test.go:1288: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:1288
        	            				pkg/testutils/soon.go:71
        	            				pkg/util/retry/retry.go:213
        	            				pkg/testutils/soon.go:77
        	            				pkg/testutils/soon.go:56
        	            				pkg/testutils/soon.go:38
        	            				pkg/ccl/changefeedccl/helpers_test.go:1272
        	            				pkg/ccl/changefeedccl/changefeed_test.go:6407
        	            				pkg/ccl/changefeedccl/helpers_test.go:1134
        	            				pkg/ccl/changefeedccl/helpers_test.go:1169
        	Error:      	Not equal: 
        	            	expected: true
        	            	actual  : false
        	Test:       	TestChangefeedContinuousTelemetryOnTermination/cloudstorage
    --- FAIL: TestChangefeedContinuousTelemetryOnTermination/cloudstorage (18.16s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedContinuousTelemetryOnTermination.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48998

I re-ran `./dev test pkg/ccl/changefeedccl --filter TestChangefeedContinuousTelemetryOnTermination -v --race --stress` and got an error.

It failed with Kafka so it's not specific to cloudstorage

```
=== RUN   TestChangefeedContinuousTelemetryOnTermination
    test_log_scope.go:165: test logs captured to: /Users/keithchow/go/src/github.com/cockroachdb/cockroach/y/_tmp/94b2d748161153a63f0302544a0eeb9e/logTestChangefeedContinuousTelemetryOnTermination291829961
    test_log_scope.go:76: use -show-logs to present logs inline
=== RUN   TestChangefeedContinuousTelemetryOnTermination/kafka
    helpers_test.go:1038: making server as secondary tenant
    helpers_test.go:1117: making kafka feed factory
    helpers_test.go:1435: read message {{{1744051476259344000 changefeed_emitted_bytes} ‹CREATE CHANGEFEED FOR TABLE foo INTO 'kafka://does.not.matter/'› kafka 1 no yes  1061621189851578369} 23 1 86400000000000 false}
    helpers_test.go:1435: read message {{{1744051477088795000 changefeed_emitted_bytes} ‹CREATE CHANGEFEED FOR TABLE foo INTO 'kafka://does.not.matter/'› kafka 1 no yes  1061621189851578369} 23 1 86400000000000 true}
    helpers_test.go:1435: read message {{{1744051477069011000 changefeed_emitted_bytes} ‹CREATE CHANGEFEED FOR TABLE foo INTO 'kafka://does.not.matter/'› kafka 1 no yes  1061621189851578369} 23 1 86400000000000 false}
    helpers_test.go:1443: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:1443
        	            				pkg/testutils/soon.go:71
        	            				pkg/util/retry/retry.go:213
        	            				pkg/testutils/soon.go:77
        	            				pkg/testutils/soon.go:56
        	            				pkg/testutils/soon.go:38
        	            				pkg/ccl/changefeedccl/helpers_test.go:1427
        	            				pkg/ccl/changefeedccl/changefeed_test.go:7221
        	            				pkg/ccl/changefeedccl/helpers_test.go:1289
        	            				pkg/ccl/changefeedccl/helpers_test.go:1324
        	Error:      	Not equal: 
        	            	expected: true
        	            	actual  : false
        	Test:       	TestChangefeedContinuousTelemetryOnTermination/kafka
=== NAME  TestChangefeedContinuousTelemetryOnTermination
    changefeed_test.go:7225: -- test log scope end --
test logs left over in: /Users/keithchow/go/src/github.com/cockroachdb/cockroach/y/_tmp/94b2d748161153a63f0302544a0eeb9e/logTestChangefeedContinuousTelemetryOnTermination291829961
--- FAIL: TestChangefeedContinuousTelemetryOnTermination (16.27s)
    --- FAIL: TestChangefeedContinuousTelemetryOnTermination/kafka (16.23s)
FAIL
I250407 18:44:41.338710 1 (gostd) testmain.go:661  [-] 1  Test //pkg/ccl/changefeedccl:changefeedccl_test exited with error code 1
```

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -7553,6 +7553,7 @@ func TestChangefeedContinuousTelemetry(t *testing.T) {
 
 type testTelemetryLogger struct {
 	telemetryLogger
+	id                      int32
 	afterIncEmittedCounters func(numMessages int, numBytes int)
 }
 
@@ -7561,6 +7562,18 @@ var _ telemetryLogger = (*testTelemetryLogger)(nil)
 func (t *testTelemetryLogger) incEmittedCounters(numMessages int, numBytes int) {
 	t.telemetryLogger.incEmittedCounters(numMessages, numBytes)
 	t.afterIncEmittedCounters(numMessages, numBytes)
+}
+
+func (t *testTelemetryLogger) maybeFlushLogs() {
+	if t.id == 1 {
+		t.telemetryLogger.maybeFlushLogs()
+	}
+}
+
+func (t *testTelemetryLogger) close() {
+	if t.id == 1 {
+		t.telemetryLogger.close()
+	}
 }
 
 func TestChangefeedContinuousTelemetryOnTermination(t *testing.T) {
@@ -7583,6 +7596,7 @@ func TestChangefeedContinuousTelemetryOnTermination(t *testing.T) {
 			}
 			return nil
 		}
+		var numPeriodicTelemetryLogger atomic.Int32
 		// Synchronization to prevent a race between the changefeed closing
 		// and the telemetry logger getting emitted counts after messages
 		// have been emitted to the sink.
@@ -7595,6 +7609,7 @@ func TestChangefeedContinuousTelemetryOnTermination(t *testing.T) {
 						seen.Store(true)
 					}
 				},
+				id: numPeriodicTelemetryLogger.Add(1),
 			}
 		}
 
@@ -7623,7 +7638,14 @@ func TestChangefeedContinuousTelemetryOnTermination(t *testing.T) {
 
 		// Close the changefeed and ensure logs were created after closing.
 		require.NoError(t, foo.Close())
-		verifyLogsWithEmittedBytesAndMessages(t, jobID, afterFirstLog.UnixNano(), interval.Nanoseconds(), true /* closing */)
+
+		if numPeriodicTelemetryLogger.Load() > 1 {
+			t.Log("transient error")
+		}
+
+		verifyLogsWithEmittedBytesAndMessages(
+			t, jobID, afterFirstLog.UnixNano(), interval.Nanoseconds(), true, /* closing */
+		)
 	}
 
 	cdcTest(t, testFn)

